jwplatform/upload.py

```diff
@@ -0,0 +1,137 @@
+        file_size = common.get_file_size(self.file)
+        part_count = file_size // self.target_part_size + 1
+
+        if part_count > 1000:
+            raise ValueError(f"The given file cannot be divided into more than 1000 parts. Please try increasing the "
+                             f"target part size.")
+
+        # Get the part links
+        upload_links = self._get_pre_signed_part_links(part_count)
+
+        # Upload the parts
+        for part_number in range(1, part_count + 1):
+            bytes_chunk = self.file.read(self.target_part_size)
+            if part_number < part_count and len(bytes_chunk) != self.target_part_size:
+                raise IOError("Failed to read enough bytes")
+            retry_count = 0
+            while retry_count < self.upload_retry_count:
+                try:
+                    self._upload_part(bytes_chunk, part_number, upload_links)
+                    break
+                except (DataIntegrityError, HTTPError):
+                    self.logger.warning(f"Encountered error upload part {part_number} of {part_count} for file {filename}."
+                                        f" Retrying.")
+                    retry_count = retry_count + 1
+
+            if retry_count >= self.upload_retry_count:
+                raise MaxRetriesExceededError(f"Max retries ({self.upload_retry_count}) exceeded while uploading part"
+                                              f" {part_number} of {part_count} for file {filename}.")
+
+        # Mark upload as complete
+        self._mark_upload_completion()
+
+    def _upload_part(self, bytes_chunk, part_number, upload_links):
+        # Add a S3 server-side checksum validation too if possible.
+        computed_hash = self._compute_part_hash(bytes_chunk)
+
+        # Check if the file has already been uploaded and the hash matches. Return immediately without doing anything
+        # if the hash matches.
+        upload_hash = upload_links[part_number - 1]["etag"] if "etag" in upload_links[part_number - 1] else None
+        if upload_hash:
+            if repr(upload_hash) == repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+                return
+
+        upload_link = upload_links[part_number - 1]["upload_link"] if "upload_link" in upload_links[part_number - 1] \
+            else None
+        if not upload_link:
+            raise Exception(f"Invalid upload link for part {part_number}.")
+
+        resp = requests.put(upload_links[part_number - 1]["upload_link"], data=bytes_chunk)
+        resp.raise_for_status()
+
+        returned_hash = resp.headers['ETag']
+        if repr(returned_hash) != repr(f"\"{computed_hash}\""):  # The returned hash is surrounded by '"' character
+            raise DataIntegrityError("The hash of the uploaded file does not match with the hash on the server.")
+        self.logger.info(f"Successfully uploaded part {part_number} for upload id {self.upload_id}")
```

Comment:
This is a pretty minor point but this might be better as debug instead of info. I'm not sure this level of granularity is needed by default.